### PR TITLE
core: build zones of kotlin infra from itself only (partially-filled)

### PR DESCRIPTION
**Pull request by bougue-pe**
Also:
* remove unused switchToZone and zoneMap in RawInfraAdapter
* stabilize detector sorting used for zone's name

Note:
* keep zone() method in Builder to let Adapter orchestrate and pass map of track sections to detectors
* raises warnings on 5 zones with <2 detectors on Germany (import edge-cases).
  ```
   4   │ [16:05:56,478] [WARN]        [RawInfraBuilder] Invalid detection zone delimited by detector(s) [573418174] and containing node(s) [691570452]
   5   │ [16:05:56,490] [WARN]        [RawInfraBuilder] Invalid detection zone delimited by detector(s) [1141072995] and containing node(s) [3233572434]
   6   │ [16:05:56,502] [WARN]        [RawInfraBuilder] Invalid detection zone delimited by detector(s) [3923371281] and containing node(s) [1398993150]
   7   │ [16:05:56,509] [WARN]        [RawInfraBuilder] Invalid detection zone delimited by detector(s) [] and containing node(s) [7663005738, 7663005739]
   8   │ [16:05:56,509] [WARN]        [RawInfraBuilder] Invalid detection zone delimited by detector(s) [] and containing node(s) [8618898998, 7997895383]
  ```

Final part of partial issue https://github.com/osrd-project/osrd-confidential/issues/307
🔍 Better review by commit

### CODE REVIEW

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt [hunk 1]
```diff
@@ -87,11 +86,22 @@ class RawInfraFromRjsBuilderImpl : RawInfraBuilder {
         return trackSectionDistanceSortedChunkMap
     }
 
-    // TODO remove this accessor once useless in adapter
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: why remove this note? The TrackNodeDescriptor object isn't supposed to be public eventually, hence the comment about removing it

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 2]
```diff
@@ -278,23 +276,113 @@ val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
...
+                firstEndpoint = endpoint
+            } else {
+                uf.union(firstEndpoint.index.toInt(), endpoint.index.toInt())
+            }
+        }
+    }
+    // union by track-section without waypoint
+    for (trackSectionIdx in trackSectionPool) {
+        val waypoints =
+            trackSectionNameToDistanceSortedWaypointIdxs[trackSectionPool[trackSectionIdx].name]
```
Comment (multun) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: waypoint is the legacy name for detectors. we try not to use it anymore, as it also refers to pathfinding waypoints

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 3]
```diff
@@ -331,7 +429,13 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): SimInfraAdapter {
         val trackSectionBlockedGauges = getBlockedGauge(rjsTrack)
 
         var chunkStartOffset = Offset<TrackSection>(0.meters)
-        val chunkEndOffsets = trackSectionNameToChunkEndOffsets[rjsTrack.id] ?: mutableSetOf()
+        val trackSectionDistanceSortedWaypointIdxs =
+            trackSectionNameToDistanceSortedWaypointIdxs[rjsTrack.id] ?: TreeMap()
+        // ignore 0 offset that is a chunk's start
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: The comment is a little unclear, maybe something like this instead?

```suggestion
        // split chunks on any detector that isn't placed at offset=0
```

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 4]
```diff
@@ -278,23 +276,113 @@ val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
...
+        if (waypoints.isNullOrEmpty()) {
+            val startEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.START)
+            val endEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.END)
+            uf.union(startEndpoint.index.toInt(), endEndpoint.index.toInt())
+        }
+    }
+
+    // now that union-find is complete, create detection zones
+    val zoneRootToNodesAndDirDetectors =
+        mutableMapOf<Int, Pair<MutableSet<TrackNodeId>, MutableSet<DirDetectorId>>>()
```
Comment (multun) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: The pair of track node and dirdetectorid looks a lot like a the internal zone descriptor. I believe the code would look nicer if the builder API allowed adding track nodes to zones. The track node builder API would then be as follows:

```kotlin
fun zone(): ZoneId
fun zoneAddNode(ZoneId, TrackNodeId)
fun zoneAddBound(ZoneId, DirDetectorId)
```

In this function, you can then do:
```suggestion
val rootToZoneMap = mutableMapOf<Int, ZoneId>()
```

**Review by eckter — APPROVED** [review]
Just a minor comment, otherwise LGTM

File: core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt [hunk 5]
```diff
@@ -54,7 +54,7 @@ class BlockAvailabilityTests {
             "zone.[det.center.1:INCREASING, det.center.2:DECREASING]",
             "zone.[det.center.2:INCREASING, det.center.3:DECREASING]",
             "zone.[det.b1.nf:DECREASING, det.b2.nf:DECREASING, det.center.3:INCREASING]",
-            "zone.[det.b1.nf:INCREASING, bf.b1:DECREASING]"
+            "zone.[bf.b1:DECREASING, det.b1.nf:INCREASING]"
```
Comment (multun) [inline core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt:57]: do you know why this change is required?

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt [hunk 6]
```diff
@@ -259,15 +258,8 @@ interface RestrictedRawInfraBuilder {
     fun linkZones(detector: DetectorId, zoneA: ZoneId, zoneB: ZoneId)
 
     fun setNextZone(detector: DirDetectorId, zone: ZoneId)
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:255]: as described below, the zone method wasn't designed to create a complete zone from the get go: you create a zone handle, and then use setNextZone to define its bounds. I believe the builder API can be simplified to make the job simpler

**Review by multun — CHANGES_REQUESTED** [review]

**Review by Erashin — APPROVED** [review]
Works for me.

File: core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt (hunk 5 above)
Comment (bougue-pe) [reply core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/BlockAvailabilityTests.kt:57]: The sorting for naming is changed in this commit: https://github.com/osrd-project/osrd/pull/6921/commits/a5da254c30e3038dd14b0c361d7ff5d340593edd#diff-5d11fb62d051b79ae994acaff5b58f193131909011f76ad3547e4c777e4dfc85R389
This change was done to stabilize names and ease comparison.

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 2 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: OK, I kept `waypoint` to wrap detectors and buffer-stops.
Considering that, if you confirm that `detector` is better (confirmed ✅), I'm OK for the change (I don't have a 3rd alternative in mind).

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt (hunk 1 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: This note was removed after choosing (debatable) to put the main part of the process in adapter instead of builder.
This requires read access to nodes [here](https://github.com/osrd-project/osrd/pull/6921/commits/30286b9651e73cea7e4efd1cf6b641499f2326c4#diff-0335c3a24d409feef2e7c3f5f3173cc4d08928c658e17faa8a0e71384922e6d3R303) (did not find easy read-only api for StaticPool).
If after discussion this requirement is dropped, then I agree that this TODO must be back.

An alternative is to do specific "read-accessors" on Builder: `getNodeIterator()` and `getNodePorts(TrackNodeId)` (node name may stay internal depending on other refactor).

The debate on moving the process/changing builder's API will probably be [here](https://github.com/osrd-project/osrd/pull/6921#discussion_r1543028357).

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 3 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Done (as discussed): https://github.com/osrd-project/osrd/pull/6921/commits/754865f7833fe40281a8789e8f79108f55b6a2dc#diff-0335c3a24d409feef2e7c3f5f3173cc4d08928c658e17faa8a0e71384922e6d3R424

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt (hunk 1 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: After discussion, the API is separated and the note is back: https://github.com/osrd-project/osrd/pull/6921/commits/754865f7833fe40281a8789e8f79108f55b6a2dc#diff-439a89c8361baafe08032d2f75202a6870d30ddea85b1f2c8e78754546798179R96

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 2 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Done in https://github.com/osrd-project/osrd/pull/6921/commits/0f93e12e301ca1b649b1b10a40fcd271961674f6

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 4 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Done in https://github.com/osrd-project/osrd/pull/6921/commits/754865f7833fe40281a8789e8f79108f55b6a2dc
Which also triggers https://github.com/osrd-project/osrd/pull/6921/commits/997b08551d1ee2d8032d9d3f279ccde5b11daf78 to allow slightly different result for zones.

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt (hunk 6 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:255]: See the conversation https://github.com/osrd-project/osrd/pull/6921#discussion_r1543018849

**Review by multun — APPROVED** [review]
